suite/tests/api/ir_x86.c

```diff
@@ -787,85 +789,133 @@ test_instr_encode_and_decode(void *dc, instr_t *instr, uint len_expect,
     instr_destroy(dc, decin);
 }
 
+void
+get_instr_src_and_dst_registers(instr_t *instr, uint max_num_regs, byte *used_src_reg_map,
+                                byte *used_dst_reg_map)
+{
+    memset(used_src_reg_map, 0, max_num_regs);
+    memset(used_dst_reg_map, 0, max_num_regs);
+    uint original_num_dsts = (uint)instr_num_dsts(instr);
+    for (uint dst_index = 0; dst_index < original_num_dsts; ++dst_index) {
+        opnd_t dst_opnd = instr_get_dst(instr, dst_index);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(dst_opnd);
+        if (opnd_is_memory_reference(dst_opnd)) {
+            for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+                // TODO i#6662: need to add virtual registers.
+                // Right now using regular reg_id_t (which holds DR_REG_ values) from
+                // opnd_api.h.
+                reg_id_t reg = opnd_get_reg_used(dst_opnd, opnd_index);
+                if (used_src_reg_map[reg] == 0)
+                    used_src_reg_map[reg] = 1;
+            }
+        } else {
+            for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+                // TODO i#6662: need to add virtual registers.
+                // Right now using regular reg_id_t (which holds DR_REG_ values) from
+                // opnd_api.h.
+                reg_id_t reg = opnd_get_reg_used(dst_opnd, opnd_index);
+                if (used_dst_reg_map[reg] == 0)
+                    used_dst_reg_map[reg] = 1;
```

Comment:
Hmm, this mostly duplicates the encoding code.

It might be useful to have a whole-instr register iterator in the general API?  Should we create that and then we can use it in both places?  Separate PR?

Reply:
Yes, they are awfully close.
A whole instruction register iterator would be nice, although we still want to distinguish between source and destination registers.
Separate PR is probably better.
Keeping it as is for now.

Reply:
Please add a comment acknowledging that this is duplicating code (and now we have to keep them in synch...).  I would say let's file an issue on a whole-instr register iterator and we can put a XXX i# comment here to eliminate the duplication.

Reply:
Done.
This is the issue: https://github.com/DynamoRIO/dynamorio/issues/6717 please double check that my understanding for the new iterator is what you had in mind, thanks!

Reply:
Yes LGTM.